### DISCUSSION TITLE
GetProvider() returns undefined / null

### DISCUSSION THREAD

**Discussion by nallibtermatrac** [nextauthjs/next-auth#4351] [Help]
### Environment

{
  "name": "t5customerportal",
  "version": "0.2.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@emotion/react": "^11.8.1",
    "@emotion/styled": "^11.8.1",
    "@mui/icons-material": "^5.4.4",
    "@mui/material": "^5.4.4",
    "@mui/x-data-grid": "^5.6.1",
    "@mui/x-data-grid-generator": "^5.6.1",
    "@mui/x-data-grid-pro": "^5.6.1",
    "axios": "^0.26.1",
    "classnames": "^2.2.6",
    "mui-datatables": "^4.1.2",
    "multer": "^1.4.4",
    "next": "12.1.0",
    "next-auth": "^4.2.1",
    "next-connect": "^0.12.2",
    "pino": "^7.9.2",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-transition-group": "^4.4.2",
    "sentry": "^0.1.2",
    "sharp": "^0.30.3"
  }
}

### Reproduction URL

https://next-auth.js.org/configuration/pages

### Describe the issue

Using the example from the website... it returns null and crashes

```
import { getProviders, signIn } from "next-auth/react"

export default function SignIn({ providers }) {
  return (
    <>
      {Object.values(providers).map((provider) => (
        <div key={provider.name}>
          <button onClick={() => signIn(provider.id)}>
            Sign in with {provider.name}
          </button>
        </div>
      ))}
    </>
  )
}

// This is the recommended way for Next.js 9.3 or newer
export async function getServerSideProps(context) {
  const providers = await getProviders()
  return {
    props: { providers },
  }
}

/*
// If older than Next.js 9.3
SignIn.getInitialProps = async () => {
  return {
    providers: await getProviders()
  }
}
*/
```

### How to reproduce

just copy and paste the example from the website

### Expected behavior

getting a customized page

**Comment by ThangHuuVu** [comment]
This is an likely an issue with `NEXTAUTH_URL`. Could you double check if you set the correct value?

**Reply by nallibtermatrac** [reply]
I am using this:
NEXTAUTH_URL=https://localhost:3000

I checked if it was the HTTP or HTTPS. No luck.
Of course I have a different one in .env.production

**Reply by ThangHuuVu** [reply]
hmmm... I can't seem to reproduce this. If you could help providing a reproduction URL, it would be super helpful to debug 🙏

**Reply by magicspon** [reply]
I'm seeing this issue as well.

https://github.com/magicspon/next-auth-custom-login-page-error

example code from https://www.youtube.com/watch?v=kB6YNYZ63fw (updated to nextauth 4)

**Reply by ThangHuuVu** [reply]
@magicspon I wasn't able to reproduce using your repo.
<img width="1365" alt="Screen Shot 2022-04-25 at 09 48 52" src="https://user-images.githubusercontent.com/31528554/165012565-05ca204b-6473-4594-865c-03a6ed8327cc.png">
Make sure you have a `.env` file similar to this:
 ```
AUTH_SECRET=secret
GOOGLE_ID="your_id"
GOOGLE_SECRET="your_secret"
```

**Reply by magicspon** [reply]
wwweird... i have those values in my `.env.local`

**Reply by magicspon** [reply]
Interesting... I had docker running (via https://craftcms.com/docs/nitro/2.x/).
The page loads as expected with docker turned off.

**Reply by magicspon** [reply]
Moving nextjs to run on port 6001 has resolved this issue for me.

**Reply by jonashaefele** [reply]
I tried a whole bunch of things after this random problem popped up for me. Literally overnight without any code changes. Maybe this helps someone:

After a while, I got a bit frustrated, so my attempts to solve it got less structured. But what eventually fixed it was:

- delete both `node_modules` and `.next` folders
- re-install dependencies
- change `.env` file formatting. for some reason I had all the vars defined like `NEXTAUTH_URL = 'http://localhost:6001';` i changed it to `NEXTAUTH_URL='http://localhost:6001'` (I removed the spaces around the `=` and removed the semicolon at the end of the line.
- fresh build

I literally don't know which of the steps it was. Clearing `.next` alone didn't do the trick. Re-installing all node dependencies and re-formatting `.env` together helped. I think some VS Code extension re-formatted my `.env` which somehow lead next-auth not being able to properly read the variable, which then led to the following error:

```
[next-auth][error][CLIENT_FETCH_ERROR]
https://next-auth.js.org/errors#client_fetch_error Only absolute URLs are supported
```

and an empty providers array.

**Reply by Abhinil-07** [reply]
How to get back the .next file, I am a beginner so bit confused

**Comment by nallibtermatrac** [comment]
An update:

I have updated React, Next and NextAuth to...
"react": "17.0.2",
"next-auth": "^4.5.0",
"next": "12.1.1",

I am still receiving the same error:
error - TypeError: Cannot convert undefined or null to object

My app:
`
<SessionProvider session={session} refetchInterval={5 * 60}>
            <ThemeProvider theme={theme_}>
                <T5Header {...pageProps}></T5Header>
                <Component {...pageProps} />
                <T5Footer {...pageProps}></T5Footer>
            </ThemeProvider>
        </SessionProvider>
`

**Comment by nallibtermatrac** [comment]
Hi,
I have noticed this https://next-auth.js.org/configuration/pages#oauth-sign-in ...
doesnt work in "pages/api" but it works in "pages".

the same code as the example is showing me the providers properly. 

I am sorry for the misunderstanding

**Reply by FemtDeveloper** [reply]
hi, can you show us the working tree? Please

**Reply by kb2232** [reply]
its hard to understand what you saying - please show directory structure

**Reply by nallibtermatrac** [reply]
Hi,
What I am trying to say is I didnt know the "getproviders code" works at the hooks area, not in the server area.

**Comment by Inkvii** [comment]
i had same problem here. When I was running it using `npm run dev` it worked fine using `NEXTAUTH_URL=http://localhost:3000`. Once i built it and ran `npm run start` I got the same error about `CLIENT_FETCH_ERROR` when trying to get providers using `await getProviders()` call.

While I was debugging the problem, it seems that even using inbuilt fetch or axios on url `NEXTAUTH_URL/api/auth/providers` would result in `ERRCONNECT` but the same get request using browser would grant me the json.

In the end i had to change `NEXTAUTH_URL` and instead of localhost i had to use address that pointed to my IP (found using ipconfig /ifconfig ). That solved it fortunately

**Reply by BenoitSafari** [reply]
Thanks that did the trick for me.

**Reply by FemtDeveloper** [reply]
Hi, that in locally. Have you tried deployed in vercel?

**Reply by pauljmillar** [reply]
Thanks, reading this helped me notice I had a typo in
`NEXTAUTH_URL=http://localhost:3000`

**Comment by serg06** [comment]
For me, the issue happens when Deployment Protections are enabled.

Clearing cookies or disabling Deployment Protections fixes it.

**Reply by lisander-lopez** [reply]
2 days of me pulling my hair out google led me to this discussion... this "Vercel Authentication" feature is what I disabled to get it working........ and to think Vercel integration is "seamless"

**Reply by AbuSantos** [reply]
Vercel Authentication, where did you implement this

**Reply by lisander-lopez** [reply]
If you are asking where to go on Vercel to change this then it is,
 Go to your project page -> Click settings tab at the top -> Click Deployment Protection on the left -> The first card should be Vercel Authentication....... if im not mistaken. I would prefer to leave it on as it does provide value in locking down all the preview deployments that get created.... Here are the docs if anything https://vercel.com/docs/security/deployment-protection/methods-to-protect-deployments/vercel-authentication

**Reply by mannerism** [reply]
Turning off "settings" > "Deployment Protection" >  "Vercel Authentication - Ensures visitors to your deployments are logged in to Vercel and are a member of your team." did the trick for me.
